story/vspc-89

ashmichheda opened the pull request:
Merged changes of ```VSPC-120``` and migrated jsp related code to Thymeleaf

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/model/impl/Space.java:
```diff
@@ -35,10 +35,13 @@
     @Access(AccessType.PROPERTY)
-    private boolean showUnpublishedLinks;
-
+    private boolean showUnpublishedLinks; // -> true/false
```
remove comment

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/model/impl/Space.java:
```diff
@@ -35,10 +35,13 @@
+    private boolean showUnpublishedLinks; // -> true/false
+    
+    @Access(AccessType.PROPERTY)
+    private boolean hideLinksToUnpublishedSpaces;
```
the this boolean should indicate if all links to this space should be hidden, not to hide links to unpublished spaces (that's what the previous boolean is for)

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/model/impl/Space.java:
```diff
@@ -167,5 +170,20 @@ public Boolean hasIncomingLinks() {
+			return;
+		}
+		this.hideLinksToUnpublishedSpaces = hideLinksToUnpublishedSpaces;
+	}
```
adjust your code style settings and format

jdamerow replied:
?

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/model/impl/SpaceLink.java:
```diff
@@ -35,7 +37,16 @@
+     * 
+     * @Access(AccessType.PROPERTY) public void setHideLinks(int hideLinks) {
+     * this.hideLinks = hideLinks; }
+     */
```
remove

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/publicview/ExhibitionSpaceController.java:
```diff
@@ -58,9 +62,13 @@ public String space(@PathVariable("id") String id, Model model) {
-                model.addAttribute("spaceLinks", spaceLinkManager.getSpaceLinkForGivenOrNullSpaceStatus(id, SpaceStatus.PUBLISHED));
+            	List<ISpaceLinkDisplay> spaceLinks = spaceLinkManager.getSpaceLinkForGivenOrNullSpaceStatus(id, SpaceStatus.PUBLISHED);
+                List<ISpaceLinkDisplay> filteredSpacelinks = filterLinks(spaceLinks);
+                model.addAttribute("spaceLinks",filteredSpacelinks);
```
DRY, move duplicate code outside of if statement

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/publicview/ExhibitionSpaceController.java:
```diff
@@ -74,4 +82,16 @@ public String space(@PathVariable("id") String id, Model model) {
     }
+
+	private List<ISpaceLinkDisplay> filterLinks(List<ISpaceLinkDisplay> spaceLinks) {
+		// Logic to filter the space links with hide links status to true
```
turn comment into proper javadoc

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/publicview/ExhibitionSpaceController.java:
```diff
@@ -74,4 +82,16 @@ public String space(@PathVariable("id") String id, Model model) {
+				linksToDisplay.add(spaceLinkDisplayObj);
+			}	
+		}
+		return linksToDisplay;
```
all of the above can be done in one line using the stream api and filter method

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/UpdateShowUnpublishedLinksController.java:
```diff
@@ -29,4 +29,17 @@ public String updateStatus(HttpServletRequest request,RedirectAttributes attribu
     }
+    
+    @RequestMapping(value="/staff/space/{spaceId}/hideSpaceLinksFromPublishedSpace", method=RequestMethod.POST)
+    public String updateStatusForPublishedSpace(HttpServletRequest request,RedirectAttributes attributes, @PathVariable("spaceId") String spaceId,
```
put in its own controller

jdamerow replied:
should have a better method name (what is status here and why published space?)

jdamerow replied:
what do you need the `HttpServletRequest` parameter for?

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/UpdateShowUnpublishedLinksController.java:
```diff
@@ -29,4 +29,17 @@ public String updateStatus(HttpServletRequest request,RedirectAttributes attribu
         return "redirect:/staff/space/{spaceId}";
     }
+    
+    @RequestMapping(value="/staff/space/{spaceId}/hideSpaceLinksFromPublishedSpace", method=RequestMethod.POST)
```
change to `.../links/incoming/edit` or something

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/UpdateShowUnpublishedLinksController.java:
```diff
@@ -29,4 +29,17 @@ public String updateStatus(HttpServletRequest request,RedirectAttributes attribu
+    
+    @RequestMapping(value="/staff/space/{spaceId}/hideSpaceLinksFromPublishedSpace", method=RequestMethod.POST)
+    public String updateStatusForPublishedSpace(HttpServletRequest request,RedirectAttributes attributes, @PathVariable("spaceId") String spaceId,
+    		@RequestParam("hideSpaceLinksPublishedSpaceParam") Boolean hideLinksToUnpublishedSpaces) {
```
how about just `show` as parameter name?

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/UpdateShowUnpublishedLinksController.java:
```diff
@@ -29,4 +29,17 @@ public String updateStatus(HttpServletRequest request,RedirectAttributes attribu
+		space.setHideLinksToUnpublishedSpaces(hideLinksToUnpublishedSpaces);
+		spaceManager.storeSpace(space, null, null);
+		attributes.addAttribute("alertType", "success");
+		attributes.addAttribute("message", "Hide all space links to this Space successfully updated!");
```
Let's make the message either "All links to this space are now hidden." or "All links to this space are now shown."

jdamerow commented on vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.jsp:
```diff
@@ -10,6 +10,7 @@
     href="<c:url value="/resources/extra/Home.css" />">
 <script>
 
+
```
there should be no changes here

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/model/impl/Space.java:
```diff
@@ -38,7 +38,9 @@
     private boolean showUnpublishedLinks;
-
+    
+    private boolean hideAllIncomingLinksToGivenSpace;
```
let's call this `hideIncomingLinks`

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/publicview/ExhibitionSpaceController.java:
```diff
@@ -58,10 +66,13 @@ public String space(@PathVariable("id") String id, Model model) {
+                
             }else {
-                model.addAttribute("spaceLinks", spaceLinkManager.getSpaceLinkForGivenOrNullSpaceStatus(id, SpaceStatus.PUBLISHED));
+            	spaceLinks = spaceLinkManager.getSpaceLinkForGivenOrNullSpaceStatus(id, SpaceStatus.PUBLISHED);                
```
formatting

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/publicview/ExhibitionSpaceController.java:
```diff
@@ -74,4 +85,16 @@ public String space(@PathVariable("id") String id, Model model) {
+		/* (non-Javadoc)
+		 * Logic to filter the space links with hide links status to true and also checks if space status Unpublished
+		 * sets default all links hidden
+		 */
```
should be proper javadoc

jdamerow replied:
actually just remove the method since it's just one line of code anyways

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/UpdateHideIncomingLinksToSpaceController.java:
```diff
@@ -0,0 +1,31 @@
+		return "redirect:/staff/space/{spaceId}";
+    }
+
+}
```
formatting!

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/UpdateHideIncomingLinksToSpaceController.java:
```diff
@@ -0,0 +1,31 @@
+	@Autowired
+    private ISpaceManager spaceManager;
+	
+	@RequestMapping(value="/staff/space/links/edit/{spaceId}/hideSpaceLinksFromSpace", method=RequestMethod.POST)
```
just make this `/staff/space/links/edit/{spaceId}/incoming`

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/UpdateHideIncomingLinksToSpaceController.java:
```diff
@@ -0,0 +1,31 @@
+		space.setHideAllIncomingLinksToGivenSpace(hideAllIncomingLinksToGivenSpace);
+		spaceManager.storeSpace(space, null, null);
+		attributes.addAttribute("alertType", "success");
+		attributes.addAttribute("message", "All links to this space are now hidden.");
```
what if the `hideAllIncomingLinksToGivenSpace` was set to false? This would not be true. Also, just call `hideAllIncomingLinksToGivenSpace` `hideLinks` or something (a lot less to type and still perfectly understandable ;)

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/publicview/ExhibitionSpaceController.java:
```diff
@@ -58,10 +64,13 @@ public String space(@PathVariable("id") String id, Model model) {
-                model.addAttribute("spaceLinks", spaceLinkManager.getSpaceLinkForGivenOrNullSpaceStatus(id, SpaceStatus.PUBLISHED));
+            	spaceLinks = spaceLinkManager.getSpaceLinkForGivenOrNullSpaceStatus(id, SpaceStatus.PUBLISHED);                
             }
+            List<ISpaceLinkDisplay> filteredSpaceLinks = spaceLinks.stream().filter(spaceLinkDisplayObj -> !spaceLinkDisplayObj.getLink().getTargetSpace().getHideIncomingLinks() || spaceLinkDisplayObj.getLink().getSourceSpace().getSpaceStatus().equals(SpaceStatus.UNPUBLISHED)).collect(Collectors.toList());
```
what is this for:
` spaceLinkDisplayObj.getLink().getSourceSpace().getSpaceStatus().equals(SpaceStatus.UNPUBLISHED)`
if `space.isShowUnpublishedLinks()` is true, the one you're filtering out should be shown, otherwise they would already be filtered out.

ashmichheda replied:
If there are 2 spaces defined say Space 1, with Space ID: ```1001``` and Space 2 with Space ID: ```1002```, and I make Space 2 as ```Unpublished``` and Space 1 is ```Published```. 
When I go to the public view page of Space 1, I am checking the incoming space links space status, i.e. I am checking whether there is an incoming link from Space 2 on Space 1 and if the status of Space 2 is ```Unpublished```, I filter those links out, and it's not shown Space 1 public view page.

jdamerow replied:
but what is this for 
`spaceLinkDisplayObj.getLink().getSourceSpace().getSpaceStatus().equals(SpaceStatus.UNPUBLISHED)`
source space is the current space, isn't it? so if that wasn't published, then it wouldn't even get to this line.

ashmichheda replied:
Yes, my bad, I should change it to ```spaceLinkDisplayObj.getLink().getSourceSpace().getTargetSpace().equals(SpaceStatus.UNPUBLISHED)```

jdamerow replied:
that is taken care of in line 66/67, isn't it?

ashmichheda replied:
But that line, I suppose you mean the below line:
```spaceLinks = spaceLinkManager.getSpaceLinkForGivenOrNullSpaceStatus(id, SpaceStatus.PUBLISHED);```
Here, ```PUBLISHED``` status is being passed, but I am checking if the target space is ```UNPUBLISHED``` by default hide all the links.

jdamerow replied:
No, I mean:
```
if(space.isShowUnpublishedLinks()) {
               spaceLinks = spaceLinkManager.getLinkDisplays(id);
```
Only if a space has `showUnpublishedLinks` set to false, it should hide all links to unpublished spaces if I'm not mistaken. By adding another check for the status of the linked space you're overriding what this code is doing.

ashmichheda replied:
No it's should not override the code, because in the stream logic, I check if my flag variable ```hideIncomingLinks``` is set to ```True``` or ```False```. 
The existing code:
```if(space.isShowUnpublishedLinks()) {```
               ```spaceLinks = spaceLinkManager.getLinkDisplays(id);```
checks if other space linked to the given space is unpublished, if yes it shows all the links, and for that I do another check in the stream logic to check if ```hideIncomingLinks``` is set to ```True``` or ```False```.

ashmichheda replied:
Shall I re-open PR?

jdamerow replied:
your stream logic is to collect all links where both of these equal false, right?
`!spaceLinkDisplayObj.getLink().getTargetSpace().getHideIncomingLinks() || spaceLinkDisplayObj.getLink().getTargetSpace().getSpaceStatus().equals(SpaceStatus.UNPUBLISHED)`

so, if `spaceLinks` has 1 link (L1) to an unpublished space (S1) and 1 link (L2) to a published space (S2),
if S1.hideIncomingLink = False it would be shown
if S2.hideIncomingLink = False it would not be shown because that would be equivalent to 
`false || false` which is `false`, so it wouldn't be shown
right? Or am I missing something? why do you need the second condition (`spaceLinkDisplayObj.getLink().getTargetSpace().getSpaceStatus().equals(SpaceStatus.UNPUBLISHED)`)?

ashmichheda replied:
I am using the second condition because, by default -> if space is set to ```UNPUBLISHED``` I need to hide all the links right, irrespective of the option for hide/unhide links provided that sets ```hideIncomingLinks``` to ```TRUE/FALSE```. Hence, the second condition is to check whether the space is ```UNPUBLISHED/PUBLISHED```.

jdamerow replied:
no, this is what lines 66-71 should take care of

ashmichheda replied:
So, should I remove the second condition then?

jdamerow replied:
yes, unless there is a valid reason why it should be there.

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/UpdateHideIncomingLinksToSpaceController.java:
```diff
@@ -0,0 +1,31 @@
+        space.setHideIncomingLinks(hideIncomingLinks);
+        spaceManager.storeSpace(space, null, null);
+        attributes.addAttribute("alertType", "success");
+        attributes.addAttribute("message", "All links to this space are now hidden.");
```
what if `hideIncomingLInks` is true? then this message is false

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/model/impl/SpaceLink.java:
```diff
@@ -35,7 +37,8 @@
     private ISpace targetSpace;
-
+    
+    
```
remove changes in this file since they don't seem to be doing anything

ashmichheda replied:
I am not sure about ```private ISpace targetSpace;``` whether it's being used or not, because I didn't add this code for the ticket. Should I remove it anyway?

jdamerow replied:
I only mean changes to the files that you made that don't do anything like new lines, etc.

jdamerow replied:
Please avoid changing parts of files that are no actual code changes (like this new empty lines).

jdamerow replied:
?

jdamerow replied:
?

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/publicview/ExhibitionSpaceController.java:
```diff
@@ -48,28 +52,30 @@
-                model.addAttribute("spaceLinks", spaceLinkManager.getSpaceLinkForGivenOrNullSpaceStatus(id, SpaceStatus.PUBLISHED));
+            if (space.isShowUnpublishedLinks()) {
+                spaceLinks = spaceLinkManager.getLinkDisplays(id);
+
```
remove empty line

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/UpdateHideIncomingLinksToSpaceController.java:
```diff
@@ -0,0 +1,36 @@
+        if(hideIncomingLinks) {
+            attributes.addAttribute("message", "All links to this space are now hidden.");
+        }
+        else {
```
code style

jdamerow replied:
!

jdamerow commented on vspace/src/main/webapp/WEB-INF/views/staff/spaces/add.html:
```diff
@@ -1,94 +1,94 @@
-	</form>
-</div>
-</body>
+<html layout:decorate="~{layouts/main_staff}">
```
file encoding seems to be changed; change that back plese

jdamerow commented on vspace/src/main/webapp/WEB-INF/views/staff/spaces/menu/spaces_menu.html:
```diff
@@ -1,34 +1,34 @@
+		 <ul id="spacelist" class="nav flex-column mb-2">
+		 </ul>
+	</div>
+</body>
```
wrong line endings

jdamerow reviewed (changes requested):
Please check all your files to make sure they have the correct line endings.

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/publicview/ExhibitionSpaceController.java:
```diff
@@ -48,28 +52,30 @@
-         * Below null check is added to accommodate already existing spaces with null space status
-         */
-        if(space.getSpaceStatus() == null || space.getSpaceStatus().equals(SpaceStatus.PUBLISHED)
+        List<ISpaceLinkDisplay> spaceLinks;
```
don't remove comment

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/model/impl/Space.java:
```diff
@@ -167,5 +194,18 @@ public Boolean hasIncomingLinks() {
-    
+
+    @Override
+    public boolean getHideIncomingLinks() {
```
should be called `isHideIncomingLinks`

jdamerow commented on vspace/src/main/webapp/WEB-INF/views/staff/spaces/edit.html:
```diff
@@ -1,29 +1,29 @@
+	</div>
+</body>
+</html>
+
```
did you change anything in this file? if yes, what?

jdamerow replied:
file endings are wrong (I assume that's what changed)

ashmichheda replied:
No, I didn't change anything. I got the updated code changes on merge of `VSPC-120`

jdamerow replied:
they are still wrong. please fix.

jdamerow replied:
nver mind

jdamerow reviewed (changes requested):
- why are you adding an empty space.jsp file?
- please reread the confluence page about pull request and rework this PR accordingly.

jdamerow reviewed (commented):
please fix your line ending settings!
- list.jsp
- web.xml

jdamerow reviewed (changes requested):
Please remove empty space.jsp.

jdamerow commented on vspace/src/main/webapp/WEB-INF/views/staff/user/list.jsp:
```diff
@@ -0,0 +1,60 @@
+<%@ page pageEncoding="UTF-8" %>
```
why is this file here?

ashmichheda replied:
Actually, these files are not part of my story. I merge the latest code changes of ```VSPC-120``` and then re-open PR, which is why I don't change any formatting if any. Is there any way by which I can see what line endings or files to be removed after merging another ticket's code changes?

jdamerow replied:
You're adding this file, so it cannot have been in VSPC-120. same with space.jsp.

ashmichheda replied:
Okay. So should I delete `list.jsp` and `space.jsp`. Since I am not using them?

jdamerow replied:
you shouldn't add them (which means you need to delete them from your branch).

Also, please reread the pull request guidelines. This is something you should have noticed before I reviewed it.

ashmichheda replied:
Thanks Julia, I will re-read them. But, I did a merge with `VSPC-120` and added files that come on staging, not sure how did they come up. I will be careful next time.

jdamerow commented:
Please reread the PR guidelines in confluence.

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/model/impl/Space.java:
```diff
@@ -167,5 +194,18 @@ public Boolean hasIncomingLinks() {
+        if (hideIncomingLinks == null) {
+            this.hideIncomingLinks = false;
+        }
+        this.hideIncomingLinks = hideIncomingLinks;
```
if `hideInComingLinks` is null, this line will set `this.hideInComingLinks` to null.

ashmichheda replied:
It's throwing an error, since the variable is boolean.
Shall I cast the variable as Eclipse is suggesting?
`this.hideIncomingLinks = (Boolean) null;`

jdamerow replied:
what would that accomplish? and you also would still have the same problem. You set `hideIncomingLinks` to false and then you set it back to null.

ashmichheda replied:
But when it's null, I will have to set to a default value `false`, so that it can check further in `ExhibitionSpaceController` file whether it's `true/false` right?

ashmichheda replied:
Or should I set the type of `hideIncomingLinks` as primitive `boolean`?

jdamerow replied:
> But when it's null, I will have to set to a default value `false`, so that it can check further in `ExhibitionSpaceController` file whether it's `true/false` right?

Please go line by line and check what is happening in every line. It's not what you think. ;)

jdamerow replied:
what happened to the case that `hideIncomingLinks` is null. Can this not happen anymore?

ashmichheda replied:
No, it shouldn't happen. Since `hideIncomingLinks` is a `boolean` primitive type by default it's false. So it should be fine.

jdamerow replied:
Have you tried what happens if the value is `null` in the db (for spaces that existed before this property was added)? Hibernate would probably use the setter method to set  `hideIncomingLinks` and pass `null`, which would most likely create an exception is my guess.

ashmichheda replied:
I currently tested these changes, on the existing Spaces. It seems to work fine. 
Or if I use the previous code, then the exception is getting handled when using `Boolean` type and the setter method sets `hideIncomingLinks` to `false` when it is `null`.

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/UpdateHideIncomingLinksToSpaceController.java:
```diff
@@ -0,0 +1,35 @@
+import edu.asu.diging.vspace.core.services.ISpaceManager;
+
+@Controller
+public class UpdateHideIncomingLinksToSpaceController {
```
How about `UpdateHideLinksController`?

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/UpdateHideIncomingLinksToSpaceController.java:
```diff
@@ -0,0 +1,35 @@
+    private ISpaceManager spaceManager;
+
+    @RequestMapping(value = "/staff/space/links/edit/{spaceId}/incoming", method = RequestMethod.POST)
+    public String updateStatusForPublishedSpace(RedirectAttributes attributes, @PathVariable("spaceId") String spaceId,
```
how about just `update`

jdamerow commented on vspace/src/main/webapp/WEB-INF/views/staff/spaces/spacelist.html:
```diff
@@ -1,185 +1,185 @@
+            </div>
+        </div>
+    </div>
+</body>
```
what has been changed in this file? why does it show up as all changed?

ashmichheda replied:
I fixed the line endings, because I am not sure what files endings must have changed when I merge `VSPC-120`. So I changed line endings of all the html files.

jdamerow replied:
That's what I thought, but when I downloaded the old version, it had the correct line endings.

jdamerow commented on vspace/src/main/webapp/WEB-INF/views/staff/user/list.html:
```diff
@@ -1,47 +1,47 @@
+		        
+		        </div>
+		    </div>
+	</div>
```
same here? why does it show up as completely changed? what has actually changed?

jdamerow replied:
and this one doesn't have all lines changed (the last two stay unchanged), so it can't be line endings.

jdamerow replied:
?

jdamerow commented:
also, there are conflicts

jdamerow commented:
Make it so, Jenkins.

jdamerow reviewed (changes requested):
There are still unaddressed comments.

jdamerow commented:
Also, layout should be the same as the rest.
![image](https://user-images.githubusercontent.com/8881141/105739961-b0778f00-5f06-11eb-9b53-17951592e411.png)

ashmichheda commented:
> There are still unaddressed comments.

- The extra spaces it's not showing up in `SpaceLink.java`. In lines 40, 41 there is `java-doc` comments written.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented on vspace/src/main/webapp/WEB-INF/views/staff/user/list.html:
```diff
@@ -1,47 +1,47 @@
+		        
+		        </div>
+        </div>
+    </div>
```
Still the same question, why is this file being changed?

ashmichheda replied:
I am not sure, why this file is being changed. I have replaced the current version with the version of `VSPC-120` ticket.

jdamerow replied:
It's still here

ashmichheda replied:
This file is the version from `VSPC-120` ticket. Do I have to change the line endings? or do a File -> format?

jdamerow replied:
If this were the file from 120, it wouldn't show up as changed.

ashmichheda replied:
I replaced the contents of the file from 120 to 89. Which is why I had to commit and push them. 
What should I do with this file?

jdamerow replied:
revert it if you didn't change anything. Whatever you did, you changed the file, otherwise it wouldn't show up as changed in the pr changes tab.

ashmichheda replied:
Okay. I did File -> Replace With -> Branch -> Develop version

jdamerow replied:
You pr is not to develop though.

ashmichheda replied:
Should I replace it with `HEAD REVISION` ?

jdamerow replied:
It should be the version of the branch your are making the pr to

ashmichheda replied:
So I am making the PR to vspc-120. Then I should be replacing the file to branch 120 version right?

jdamerow replied:
Yes, replace the file with the one from 120.

ashmichheda replied:
Yes, I replaced the one from VSPC-120

diging-jenkins commented:
Congratulations, build succeeded!
